java/src/main/java/com/google/appengine/tools/pipeline/impl/util/SerializationUtils.java

```diff
@@ -1,97 +1,60 @@
-//
-// Unless required by applicable law or agreed to in writing, software
-// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
-// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
-// License for the specific language governing permissions and limitations under
-// the License.
-
+/**
+ * Copyright 2024, Worklytics, Co.
+ *
+ */
 package com.google.appengine.tools.pipeline.impl.util;
 
 import java.io.ByteArrayInputStream;
 import java.io.ByteArrayOutputStream;
 import java.io.IOException;
-import java.io.InputStream;
 import java.io.ObjectInputStream;
 import java.io.ObjectOutputStream;
-import java.util.zip.Deflater;
-import java.util.zip.DeflaterOutputStream;
-import java.util.zip.Inflater;
-import java.util.zip.InflaterInputStream;
+import java.util.zip.*;
 
 /**
- * @author rudominer@google.com (Your Name Here)
+ * worklytics version of this class replaces Google version w same public interface/behavior, but replacement offers
+ * much better performance, exploits modern java behavior; eliminates need to measure headers ourselves, etc.
  *
+ * q: use JSON serialization? more standard and potentially interoperable across languages and java versions
  */
 public class SerializationUtils {
 
-  private static final int MAX_UNCOMPRESSED_BYTE_SIZE = 1000000;
-  private static final int ZLIB_COMPRESSION = 1;
-
-  private static class InternalByteArrayOutputStream extends ByteArrayOutputStream {
-
-    public InternalByteArrayOutputStream(int size) {
-      super(size);
-    }
-
-    private byte[] getInternalBuffer() {
-      return buf;
-    }
-  }
-
-  public static byte[] serialize(Object x) throws IOException {
-    InternalByteArrayOutputStream bytes = new InternalByteArrayOutputStream(512);
-    try (ObjectOutputStream out = new ObjectOutputStream(bytes)) {
-      out.writeObject(x);
-    }
-    if (bytes.size() <= MAX_UNCOMPRESSED_BYTE_SIZE) {
-      return bytes.toByteArray();
-    }
-    ByteArrayOutputStream compressedBytes = new ByteArrayOutputStream(bytes.size() / 4);
-    compressedBytes.write(0);
-    compressedBytes.write(ZLIB_COMPRESSION);
-    Deflater deflater =  new Deflater(Deflater.BEST_COMPRESSION, true);
-    try (DeflaterOutputStream out = new DeflaterOutputStream(compressedBytes, deflater)) {
-      // Use internal buffer to avoid copying it.
-      out.write(bytes.getInternalBuffer(), 0, bytes.size());
-    } finally {
-      deflater.end();
+  // err on small; test >= this actually runs faster than < this, suggesting efficiency handling compressed data
+  private static final int MAX_UNCOMPRESSED_BYTE_SIZE = 50_000;
+
+  public static byte[] serialize(Object obj) throws IOException {
+    // Serialize the object
+    ByteArrayOutputStream byteOut = new ByteArrayOutputStream();
+    try (ObjectOutputStream objectOut = new ObjectOutputStream(byteOut)) {
+      objectOut.writeObject(obj);
+    }
+    byte[] serializedData = byteOut.toByteArray();
```

Comment:
just ask for the size `byteOut.size()`, doing `toByteArray` is copying the object into a new array, so consuming x2 memory if it happens to be big.
```
if (byteOut.size() > MAX) {
// compress
} else {
// return as is
return byteou.toByteArray();
}
```